### PR TITLE
feat: add adaptive sampling, motion registration, MEDL, and e2e ADS configs

### LINKED ISSUE #318: direct v2.2.0 tracking
- [ ] Add modulated convolutions
- [ ] Make training/validation fields optional 
- [ ] Add adaptive sampling modules
- [ ] Add registration modules
- [ ] Add end-to-end sampling, reconstruction and registration modules
- [ ] Update checkpoints in hugginface with current code

### PR DESCRIPTION
**Pull request by georgeyiasemis**
## Description

This PR brings **end-to-end adaptive dynamic sampling (ADS)** and optional **motion registration** into DIRECT, together with the paper experiment layouts for the MIDL 2026 e2e ADS works.

It wires learned sampling policies and registration models through the MRI training/inference stack, ships paper configs under `projects/e2e_ads_recon` and `projects/e2e_ads_recon_reg` (each YAML has train/val plus an `inference` block), and hardens config/checkpoint loading so those projects run cleanly on current `main` (including ruff 0.16).

## What's new
...
- **Paper projects**:
  - `projects/e2e_ads_recon` — adaptive sampling + reconstruction (vSHARP / MEDL variants, frame/init2 configs).
  - `projects/e2e_ads_recon_reg` — same family with registration (vSHARP / VarNet / LOUPE / fixed-mask variants, phase and disjoint options).
- **Inference on the same YAML**: active val-4× under `inference`, with commented 6×/8× alternatives (no per-rate file explosion).

## What's changed
...
- Docs/READMEs for adaptive, registration, MEDL, and both e2e ADS projects (including CMRxRecon fully sampled cine data notes); elastic deformation via in-repo SciPy port (no `elasticdeform` dependency).
...

## User impact

- **You can train and run** the e2e ADS (and ADS+registration) paper configs from this repo; use the same `{name}.yaml` for `direct train` and `direct predict`.
- **Inference blocks** live on the training YAML; `training` / `validation` remain optional for Hub-style inference-only configs elsewhere.
- **Registration models** need a reference image in the batch (or `registration_simulate_reference` in transforms); without that, predict fails with a missing `reference_image` key.
- **Adaptive / DYNAMIC sampling** may skip batches on rejection-sampling failure during training (logged retries) instead of always aborting immediately.
- **ADS predict** in these projects is retrospective on fully sampled volumes (model still only sees ACS/init + policy-selected lines).

### CODE REVIEW

File: direct/nn/adaptive/binarizer.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+    @staticmethod
+    def forward(ctx, inputs, slope, clamp):
+        batch_size = len(inputs)
+        probs = []
+        results = []
+
+        for i in range(batch_size):
+            x = inputs[i : i + 1]
+            count = 0
```
Comment (jonasteuwen) [inline direct/nn/adaptive/binarizer.py:43]: Can you do @jit.compile? Or doesn't that work here

File: direct/nn/adaptive/binarizer.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            Slope of the sigmoid used in the straight-through backward pass.
+        clamp : bool
+            If ``True``, clamp gradients with :func:`torch.nn.functional.hardtanh`.
+        """
+        super().__init__()
+        self.slope = slope
+        self.clamp = clamp
+
+        self.fun = ThresholdSigmoidMaskFunction.apply
+
```
Comment (jonasteuwen) [inline direct/nn/adaptive/binarizer.py:96]: I prefer self._variable for class private variables

File: direct/nn/lpd/lpd_engine.py [hunk 3]
```diff
@@ -46,8 +46,6 @@ def __init__(
         )
 
     def forward_function(self, data: dict[str, Any]) -> tuple[torch.Tensor, None]:
-        data["sensitivity_map"] = self.compute_sensitivity_map(data["sensitivity_map"])
-
```
Comment (jonasteuwen) [inline direct/nn/lpd/lpd_engine.py:50]: Why is this gone?

File: direct/nn/rim/rim.py [hunk 4]
```diff
@@ -296,7 +296,9 @@ def __init__(
                 multiscale_depth=initializer_multiscale,
             )
 
-        self.image_initialization = image_initialization
+        self.image_initialization = (
+            image_initialization.lower() if isinstance(image_initialization, str) else image_initialization
+        )
```
Comment (jonasteuwen) [inline direct/nn/rim/rim.py]: Hmm, feels like you should override the `__eq__` in the object and make sure str doesn't end up here.

File: direct/nn/varsplitnet/varsplitnet_engine.py [hunk 5]
```diff
@@ -64,8 +64,6 @@ def __init__(
         )
 
     def forward_function(self, data: dict[str, Any]) -> tuple[torch.Tensor, None]:
-        data["sensitivity_map"] = self.compute_sensitivity_map(data["sensitivity_map"])
-
```
Comment (jonasteuwen) [inline direct/nn/varsplitnet/varsplitnet_engine.py:68]: Why did this go

File: direct/registration/elastic_deformation.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    ``elasticdeform`` wheels may be incompatible with NumPy 2.x.
+
+    Returns
+    -------
+    list[np.ndarray]
+        Deformed images returned by ``elasticdeform``.
+    """
+    try:
+        from elasticdeform import deform_random_grid
+    except ImportError as exc:
```
Comment (jonasteuwen) [inline direct/registration/elastic_deformation.py]: Isn't this a function that can be forked/vendored?

File: direct/environment.py [hunk 7]
```diff
@@ -329,24 +329,27 @@ def initialize_models_from_config(
...
     # TODO(jt): Model name is not used here.
     additional_models = {}
     for k, v in cfg.additional_models.items():
         # Remove model_name key
         curr_model = models[k]
         curr_model_cfg = {kk: vv for kk, vv in v.items() if kk not in ["engine_name", "model_name"]}
-        additional_models[k] = curr_model(**curr_model_cfg)
+        curr_model_cfg.update(operator_kwargs)
+        # Drop keys unknown to the constructor (e.g. ModConv defaults on legacy modules).
+        additional_models[k] = curr_model(**filter_arguments_by_signature(curr_model, curr_model_cfg))
```
Comment (jonasteuwen) [inline direct/environment.py]: Can this cause silent bugs?

File: docker/Dockerfile [hunk 8]
```diff
@@ -1,6 +1,6 @@
 ARG CUDA="11.3.0"
 ARG PYTORCH="1.10"
-ARG PYTHON="3.8"
+ARG PYTHON="3.12"
 
 # TODO: conda installs its own version of cuda
 FROM nvidia/cuda:${CUDA}-devel-ubuntu18.04
```
Comment (jonasteuwen) [inline docker/Dockerfile]: Probably need to boost this too

File: pyproject.toml [hunk 9]
```diff
@@ -40,12 +40,14 @@ dependencies = [
...
+    "SimpleITK>=2.4",
+    "elasticdeform>=0.5",
 ]
 
 [dependency-groups]
 dev = [
     "pytest>=9",
-    "ruff>=0.16.1",
+    "ruff>=0.14",
     "ty>=0.0.38",
```
Comment (jonasteuwen) [inline pyproject.toml:51]: I would keep it to 0.16.1 or you run into issues.

**Review by jonasteuwen — COMMENTED** [review]
Please look at my comments

File: direct/nn/lpd/lpd_engine.py (hunk 3 above)
Comment (georgeyiasemis) [reply direct/nn/lpd/lpd_engine.py:50]: This was a bug. Sens estimation is already happening in main engine

File: direct/nn/varsplitnet/varsplitnet_engine.py (hunk 5 above)
Comment (georgeyiasemis) [reply direct/nn/varsplitnet/varsplitnet_engine.py:68]: This was a bug. Sens estimation is already happening in main engine

File: direct/nn/adaptive/binarizer.py (hunk 1 above)
Comment (georgeyiasemis) [reply direct/nn/adaptive/binarizer.py:43]: No because of the rejection sampling

File: direct/nn/adaptive/binarizer.py (hunk 2 above)
Comment (georgeyiasemis) [reply direct/nn/adaptive/binarizer.py:96]: done

File: direct/registration/elastic_deformation.py (hunk 6 above)
Comment (georgeyiasemis) [reply direct/registration/elastic_deformation.py]: done and verified matching with elasticdeform method

File: direct/environment.py (hunk 7 above)
Comment (georgeyiasemis) [reply direct/environment.py]: we are logging if something is dropped - so probs not

File: pyproject.toml (hunk 9 above)
Comment (georgeyiasemis) [reply pyproject.toml]: ok! done

**Review by jonasteuwen — APPROVED** [review]
LGTM
